partner_ref_unique/models/res_partner.py

```diff
@@ -0,0 +1,32 @@
+# Copyright 2016 Antonio Espinosa
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, api, models
+from odoo.exceptions import ValidationError
+
+
+class ResPartner(models.Model):
+    _inherit = "res.partner"
+
+    @api.multi
+    @api.constrains('ref', 'is_company', 'company_id')
+    def _check_ref(self):
+        for partner in self:
+            mode = partner.company_id.partner_ref_unique
+            if (partner.ref and (
+                    mode == 'all' or
+                    (mode == 'companies' and partner.is_company))):
+                domain = [
+                    ('id', '!=', partner.id),
+                    ('ref', '=', partner.ref),
+                ]
+                if mode == 'companies':
+                    domain.append(('is_company', '=', True))
+                other = self.search(domain)
+
+                # active_test is False when called from
```

Comment:
Might there not be a lot of other situations where active_test is False? In all these cases the constraint will be disabled...

Reply:
as far as I can see the odoo core will only that active_test to false in base.partner.merge.automatic.wizard, like the comment stated. only 3rd party modules could bypass this.